compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2017,755 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
+bool isSimpleSignedCompareToConst(TR::Node *secondChild, const CompareInfo& compareInfo, TR::CodeGenerator *cg)
+   {
+   if (compareInfo.isUnsignedOrUnordered)
+      return false;
 
-   node->setRegister(trgReg);
-   cg->decReferenceCount(firstChild);
-   cg->decReferenceCount(secondChild);
-   return trgReg;
+   if (compareInfo.type == TR::Int64 && !cg->comp()->target().is64Bit())
+      return false;
+
+   return secondChild->getOpCode().isLoadConst();
    }
 
-void OMR::Power::TreeEvaluator::genBranchSequence(TR::Node* node,
-   TR::Register* src1Reg, TR::Register* src2Reg, TR::Register* trgReg,
-   TR::InstOpCode::Mnemonic branchOpCode, TR::InstOpCode::Mnemonic cmpOpCode, TR::CodeGenerator* cg)
+int32_t getTypeBitLength(TR::DataTypes dt)
    {
-   TR::Register    *condReg = cg->allocateRegister(TR_CCR);
-   TR::PPCControlFlowInstruction *cfop = (TR::PPCControlFlowInstruction *)
-      generateControlFlowInstruction(cg, TR::InstOpCode::setbool, node);
-   cfop->addTargetRegister(condReg);
-   cfop->addTargetRegister(trgReg);
-   cfop->addSourceRegister(src1Reg);
-   cfop->addSourceRegister(src2Reg);
-   cfop->setOpCode2Value(branchOpCode);
-   cfop->setCmpOpValue(cmpOpCode);
-   cg->stopUsingRegister(condReg);
+   switch (dt)
+      {
+      case TR::Int8:
+         return 8;
+      case TR::Int16:
+         return 16;
+      case TR::Int32:
+         return 32;
+      case TR::Int64:
+         return 64;
+      default:
+         TR_ASSERT_FATAL(false, "Invalid data type %s for getTypeBitLength", TR::DataType::getName(dt));
+      }
    }
 
+bool isTypeSubRegister(TR::DataTypes dt, TR::CodeGenerator *cg)
+   {
+   switch (dt)
+      {
+      case TR::Int8:
+      case TR::Int16:
+         return true;
+      case TR::Int32:
+         return cg->comp()->target().is64Bit();
+      case TR::Int64:
+      case TR::Address:
+         return false;
+      default:
+         TR_ASSERT_FATAL(false, "Invalid data type %s for isTypeSubRegister", TR::DataType::getName(dt));
+      }
+   }
 
-// also handles bcmplt, scmplt
-TR::Register *OMR::Power::TreeEvaluator::icmpltEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+TR::Register *intOrderEvaluator(TR::Node *node, const CompareInfo& compareInfo, TR::CodeGenerator *cg)
    {
-   TR::Node     *firstChild   = node->getFirstChild();
-   TR::Node     *secondChild  = node->getSecondChild();
-   TR::Register *src1Reg      = cg->evaluate(firstChild);
-   TR::Register *trgReg    = cg->allocateRegister();
-   if (secondChild->getOpCode().isLoadConst() &&
-       secondChild->getRegister() == NULL)
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+
+   TR::Register *trgReg = cg->allocateRegister();
+
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P10))
+      {
+      generateCompareSetBoolean(trgReg, node, firstChild, secondChild, compareInfo, cg);
+      }
+   else if (isSimpleSignedCompareToConst(secondChild, compareInfo, cg))
```

Comment:
Hmm, you're right that this _is_ what we used to do. This does provide a bit of a tradeoff, as failing to use this path when a constant is already in a register will result in a branch instruction being used (which can be avoided given knowledge of the value's sign), while using this path can result in the same constant being loaded into a register many times which is especially problematic on 64-bit.

Maybe the correct solution here is to adjust how this optimization is being handled in the first place? Really, we can apply this optimization (at least partially) whenever either `secondChild->isNonNegative()` or `secondChild->isNonPositive()` is true. For cases where we don't use `addConstantToInteger` or `addConstantToLong` (because the second child is non-constant or is referenced multiple times), we may need an extra `addi rX, rX, -1` to handle the `(x - y - 1) < 0` comparison that sometimes appears, but that's likely a reasonable tradeoff. What are your thoughts on this?

Reply:
It seems like using branch-less code would be beneficial. But maybe we can do it in another PR and add a check for the constant value already being in register for now?

Reply:
I don't like the idea of delaying a small change like this without adequate reason. I'm not in any particular rush to get this PR in, this should be a rather simple change, this should resolve your concerns without making an unnecessary tradeoff, and making an improvement like this would be well within the scope of this PR given that this is effectively a ground-up rewrite of these evaluators. I'd really prefer to get this sorted out now rather than delay it as a later improvement, which could easily result in this being forgotten about.

From your reply, it seems like you're in favour of my proposed solution. For now, I'll go ahead and implement that in addition to dealing with the other issues you pointed out.

Reply:
Sure, we can make it part of this PR.